### ISSUE TITLE
goroutine leak when using websocket transport.

### ISSUE THREAD

**Issue by lyw0149** [improbable-eng/grpc-web#664]
**Versions of relevant software used**
grpcweb 0.12.0

**What happened**
goroutine leak when using websocket transport

**What you expected to happen**
Using xhr transport,  goroutines decrease back after connection terminated. (using pprof)

**How to reproduce it (as minimally and precisely as possible)**:
using pprof, After the stream connection is terminated, goroutines do not decrease back.
So every time reconnect, goroutine constantly increases.

**Full logs to relevant components**

**Anything else we need to know**

**Comment by johanbrandhorst** [comment]
Hi Yangwoo, thanks for your issue report! I'm a little confused, I don't know what your logs are showing. How did you determine that it was leaking a goroutine? Can you say where it was created?

As for your question, I don't understand what you mean by "open multiple browser windows". If you mean that you want to make requests from several browser sessions, you shouldn't need websockets for that.

**Comment by lyw0149** [comment]
I found goroutine leak using [pprof](https://github.com/google/pprof).
When i use xhr transport, goroutine count is increased, but decrease when disconnected.
But, in websocket transport, goroutine's count don't decrease when disconnected.

**Comment by lyw0149** [comment]
If open multiple browser windows via xhr transport and use one stream in it, is it correct that can only open up to six windows?(chrome)

**Comment by johanbrandhorst** [comment]
I don't think so. I think there was a limitation a while ago about having a most 6 websocket streams in Firefox, but I don't know of any limitation in the number of Fetch or XHR requests you can make.

> But, in websocket transport, goroutine's count don't decrease when disconnected.

Is this also true if you make 10 connections and then disconnect? How long after the disconnection have you sampled?

**Comment by lyw0149** [comment]
It increases as much as reconnect. And It's been an hour from last disconnection. 
Actually I think this issue is related with stream.Context().Done() (at grpc server) isn't trigger when websocket disconnected.
![image](https://user-images.githubusercontent.com/10323205/75609124-55e48000-5b49-11ea-9120-66c744069532.png)

**Comment by lyw0149** [comment]
> I don't think so. I think there was a limitation a while ago about having a most 6 websocket streams in Firefox, but I don't know of any limitation in the number of Fetch or XHR requests you can make.

When i use stream with xhr transport, one stream makes one connection. So I can use only 6 streams in chrome browser. Am I right?

**Comment by johanbrandhorst** [comment]
> > I don't think so. I think there was a limitation a while ago about having a most 6 websocket streams in Firefox, but I don't know of any limitation in the number of Fetch or XHR requests you can make.
> 
> When i use stream with xhr transport, one stream makes one connection. So I can use only 6 streams in chrome browser. Am I right?

I don't think so. Try it?

**Comment by johanbrandhorst** [comment]
> It increases as much as reconnect. And It's been an hour from last disconnection.
> Actually I think this issue is related with stream.Context().Done() (at grpc server) isn't trigger when websocket disconnected.
> ![image](https://user-images.githubusercontent.com/10323205/75609124-55e48000-5b49-11ea-9120-66c744069532.png)
> 
> Thanks for your fast replay. :) I like this project.

It would be great if you could do some more debugging to see where these goroutines are blocking.

**Comment by lyw0149** [comment]
> I don't think so. Try it?
![image](https://user-images.githubusercontent.com/10323205/75609564-0902a880-5b4d-11ea-80d6-c2629e2dc2cd.png)
I tried 4 streams with xhr transport in one window. The 4 connection to these 4 streams is still maintained. So if open more same new window, some streams of new window will be unavailable.

**Comment by lyw0149** [comment]
> > It increases as much as reconnect. And It's been an hour from last disconnection.
> > Actually I think this issue is related with stream.Context().Done() (at grpc server) isn't trigger when websocket disconnected.
> > ![image](https://user-images.githubusercontent.com/10323205/75609124-55e48000-5b49-11ea-9120-66c744069532.png)
> > Thanks for your fast replay. :) I like this project.
> 
> It would be great if you could do some more debugging to see where these goroutines are blocking.

Ok. I'll try it.

**Comment by lyw0149** [comment]
I think this issue is related with #663.
